comment/utils.py

```diff
@@ -206,3 +207,24 @@ def get_username_for_comment(comment):
             return comment.email.split('@')[0]
         return settings.COMMENT_ANONYMOUS_USERNAME
     return comment.user.username
+
+
+def validate_order():
+
+    @lru_cache()
+    def get_allowed_orders():
+        ALLOWED_ORDERS = ['reaction__likes', 'reaction__dislikes', 'posted']
+        # adds support for descending order as well
+        ALLOWED_ORDERS.extend(list(map(lambda a: '-' + a, ALLOWED_ORDERS)))
+        # django allows this to support random order
+        ALLOWED_ORDERS.append('?')
+        return ALLOWED_ORDERS
+
+    allowed_orders = get_allowed_orders()
+    prefered_order = settings.COMMENT_ORDER_BY
+    if prefered_order in allowed_orders:
+        return prefered_order
+
+    raise ImproperlyConfigured(
+            ErrorMessage.INVALID_ORDER_ARGUMENT.format(order=prefered_order, allowed_orders=allowed_orders)
+        )
```

Comment:
```suggestion
@lru_cache()
def get_allowed_orders():
    allowed_orders = ['reaction__likes', 'reaction__dislikes', 'posted']
    # adds support for descending order as well
    allowed_orders.extend(list(map(lambda a: '-' + a, allowed_orders)))
    # django allows this to support random order
    allowed_orders.append('?')
    return allowed_orders


def validate_order():
    preferred_orders = settings.COMMENT_ORDER_BY
    unique_values = set(map(lambda a: a.replace('-', ''), preferred_orders))
    if len(unique_values) != len(preferred_orders):
        duplicated_orders = list(set(preferred_orders) - unique_values)
        raise ImproperlyConfigured(
            ErrorMessage.DUPLICATE_ORDER_VALUE.format(
                duplicates=duplicated_orders, order=duplicated_orders[0].replace('-', '')
            )
        )

    allowed_orders = get_allowed_orders()
    for preferred_order in preferred_orders:
        if preferred_order not in allowed_orders:
            raise ImproperlyConfigured(
                ErrorMessage.INVALID_ORDER_ARGUMENT.format(order=preferred_order, allowed_orders=allowed_orders)
            )

    return preferred_orders
```

This is what came up with me. I tested and it's working fine with order list. Feel free to refactor it and please pay attention to this:
- Please use lowercase for variable name `ALLOWED_ORDER`.
- Please don't use nested function. I will show you my reason in the following comment.
- I used this validation message:
```python
    DUPLICATE_ORDER_VALUE = _((
        'Comment app: COMMENT_ORDER_BY should not have duplicated values '
        'Duplicated Values: {duplicates}. Please use one value only E.g. "{order}" or "-{order}".'
    ))
```

Reply:
My reason to not use nested functions is to avoid classing/mixing scope variables. This screenshot elabroates.
![Screenshot 2021-01-23 at 09 31 41](https://user-images.githubusercontent.com/25413382/105576365-8d1bdb00-5d72-11eb-9c04-a4fe6dc865f7.png)

You can clearly see how the variables clash.

Reply:
> Please use lowercase for variable name ALLOWED_ORDER.

It is a constant, and it is advisable to use all capital letters for such cases. You may see the PEP8 guidelines https://www.python.org/dev/peps/pep-0008/#id48

>  DUPLICATE_ORDER_VALUE = _((
        'Comment app: COMMENT_ORDER_BY should not have duplicated values '
        'Duplicated Values: {duplicates}. Please use one value only E.g. "{order}" or "-{order}".'
    ))

Disallowing duplicate values seems like moral policing to me. This thing is syntactically valid and throws up some result. Just because we can't explain the result, should we disallow it?

Reply:
> My reason to not use nested functions is to avoid classing/mixing scope variables. This screenshot elabroates.

You have used a particular example. I agree that it wouldn't have been a great practice in that particular case. 

In the current scenario, I don't think any variables from outer scopes were used inside the nested function. Also, in this case, it also prevents exposing the nested function to any other part of the code as they don't necessarily need to do anything with it.

Trying to extrapolate and generalize the issue caused in the screenshot to every example is an overreaction and maybe a classic of [faulty generalization](https://en.wikipedia.org/wiki/Faulty_generalization).

If it were that true, surely they might have not used this in python codebase or mentioned some sort of warning to **never** use nested functions in maybe at lease one of the PEPs.

Reply:
> It is a constant, and it is advisable to use all capital letters for such cases. You may see the PEP8 guidelines https://www.python.org/dev/peps/pep-0008/#id48

This is valid for module variables/constants but it is not for functions and methods scope variables.

PEP8 statement:
> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include MAX_OVERFLOW and TOTAL.

---------------
> Disallowing duplicate values seems like moral policing to me. This thing is syntactically valid and throws up some result. Just because we can't explain the result, should we disallow it?

I didn't get what you refer to!

Reply:
> I didn't get what you refer to!

As mentioned in your suggested code, you're trying to disallow duplicated values passed to the list and hence the error message regarding duplication.

> This is valid for module variables/constants but it is not for functions and methods scope variables.

Okay, so moving this to a module level constant should be a better thing. It might also settle the debate on the use of nested functions as of now.

Reply:
> You have used a particular example. I agree that it wouldn't have been a great practice in that particular case.
> 
> In the current scenario, I don't think any variables from outer scopes were used inside the nested function. Also, in this case, it also prevents exposing the nested function to any other part of the code as they don't necessarily need to do anything with it.

My statement is general, it's not related to a particular example. using nested function will most likely cause a variables clashing. Why would I need to prevent exposing such a method? What's wrong with that? 🤔 

> Trying to extrapolate and generalize the issue caused in the screenshot to every example is an overreaction and maybe a classic of [faulty generalization](https://en.wikipedia.org/wiki/Faulty_generalization).

Take it easy. Your language seems to be aggressive here.

> If it were that true, surely they might have not used this in python codebase or mentioned some sort of warning to **never** use nested functions in maybe at lease one of the PEPs.

Also you are mixing things up, nested functions are needed for some and special use cases, why would PEP inform us to not use them?!!

Reply:
> As mentioned in your suggested code, you're trying to disallow duplicated values passed to the list and hence the error message regarding duplication.

Yes duplication should not be allowed, right?

> Okay, so moving this to a module level constant should be a better thing. It might also settle the debate on the use of nested functions as of now.

I didn't say, move the variable to the module level!! I said use lowercase name for variables in the function's scope.
Uppercase names are used for module variables as per PEP8 and THIS IS NOT OUR CASE.

Reply:
> Take it easy. Your language seems to be aggressive here.

I'm sorry if it came out that way, I was just trying out to point out something that seemed as a potential [logical fallacy](https://en.wikipedia.org/wiki/Formal_fallacy) to me in your argument.

> My statement is general, it's not related to a particular example. using nested function will most likely cause a variables clashing.

In the example that you show in the screenshot the variable `request` is reused from an external scope and I agree it is not a good practice.
But in the function below, I don't see it as using any external variable, so how is it clashing with anything else. 
```python
lru_cache()
def get_allowed_orders():
    allowed_orders = ['reaction__likes', 'reaction__dislikes', 'posted']
    # adds support for descending order as well
    allowed_orders.extend(list(map(lambda a: '-' + a, allowed_orders)))
    # django allows this to support random order
    allowed_orders.append('?')
    return allowed_orders
```

>  Why would I need to prevent exposing such a method? What's wrong with that? thinking

This is what encapsulation is all about? This method is an internal part of implementation of logic of validating `order_by` parameter. Please correct if I'm wrong.

> Yes duplication should not be allowed, right?

Duplication is a valid in the query. 

```sh
$ python manage.py shell         
>>> from comment.models import Comment                                                     
>>>
>>> Comment.objects.all().order_by('posted', '-posted')                                    
<QuerySet [comment by moderator: First comment, reply by test: Reply comment, comment by admin: Second comment, comment by a: hi, comment by tester: new comment, comment by tester: hey, reply by tester: hi, reply by admin: hi, comment by admin: bla bla bals kdsjfds, reply by admin: hi]>
>>>
>>> Comment.objects.all().order_by('posted', '-posted', 'posted')                          
<QuerySet [comment by moderator: First comment, reply by test: Reply comment, comment by admin: Second comment, comment by a: hi, comment by tester: new comment, comment by tester: hey, reply by tester: hi, reply by admin: hi, comment by admin: bla bla bals kdsjfds, reply by admin: hi]>
```

> I didn't say, move the variable to the module level!! I said use lowercase name for variables in the function's scope.
Uppercase names are used for module variables as per PEP8 and THIS IS NOT OUR CASE.

I think we can move the variable to a module level constant because it is essentially a constant, and mark it as private since it may change with time.

Reply:
> I'm sorry if it came out that way, I was just trying out to point out something that seemed as a potential [logical fallacy](https://en.wikipedia.org/wiki/Formal_fallacy) to me in your argument.

No worries. Just wanted to make sure nothing disturbed you 😄 


> In the example that you show in the screenshot the variable `request` is reused from an external scope and I agree it is not a good practice.
> But in the function below, I don't see it as using any external variable, so how is it clashing with anything else.
> 
> ```python
> lru_cache()
> def get_allowed_orders():
>     allowed_orders = ['reaction__likes', 'reaction__dislikes', 'posted']
>     # adds support for descending order as well
>     allowed_orders.extend(list(map(lambda a: '-' + a, allowed_orders)))
>     # django allows this to support random order
>     allowed_orders.append('?')
>     return allowed_orders
> ```

Again, I'm generalising this issue. What if someone needed a variable name `allowed_orders` in the outer function without realizing that it's already used in the nested function? Debugging and time consuming.

I personally faced such a problem and it took me hours debugging to find that there are variable names clashing in nested scopes. My idea is to avoid these expected issues by moving the function to its own scope.

> This is what encapsulation is all about? This method is an internal part of implementation of logic of validating `order_by` parameter. Please correct if I'm wrong.

The above case voilates the encapsulation more. 😃 

> Duplication is a valid in the query.

Right, In fact you raised the duplication issue in a previous comment and I also think it's better to restrict to not repeating the value for 2 reasons:

- Duplicating the value is useless and might give unnoticeable unexpected behaviour. I am not sure which value/order will be considered here `.order_by('posted', '-posted')`.
- We are using a settings variable so it would be nice to be checked for a reasonable value and give the developer feedback on their value while development.

> I think we can move the variable to a module level constant because it is essentially a constant, and mark it as private since it may change with time.

You are using a function to manipulate this variable so I don't think the you need to use module constant variable.

Reply:
> Right, In fact you raised the duplication issue in a previous comment and I also think it's better to restrict to not repeating the value for 2 reasons:
    Duplicating the value is useless and might give unnoticeable unexpected behaviour. I am not sure which value/order will be considered here .order_by('posted', '-posted').
    We are using a settings variable so it would be nice to be checked for a reasonable value and give the developer feedback on their value while development.

Who are we to restrict them to something when `django` allows it? I agree the results can be weird but if it were not stopped by `django` maybe we should also consider allowing them.

If we want to be too opinionated(in my opinion), we can give warnings to the developer for duplicate values.

Reply:
> If we want to be too opinionated(in my opinion), we can give warnings to the developer for duplicate values.

This is all about. How to give the developer warning?!

Keep in mind that if X or Y uses a particular method or a specific lifestyle, that doesn't mean it fits your style or your need but it just fits theirs.

Reply:
> How to give the developer warning?!

Using [`RunTimeWarning`](https://docs.python.org/3/library/exceptions.html#RuntimeWarning)

> Keep in mind that if X or Y uses a particular method or a specific lifestyle, that doesn't mean it fits your style or your need but it just fits theirs.

I'm afraid I again feel this to be a generalization of my statement. Let me try to clarify a bit. 
We are building an app/library that adds functionality to `django`. So in my opinion, it don't think it would be great to restrict something that `django` allows.

Reply:
> I'm afraid I again feel this to be a generalization of my statement. Let me try to clarify a bit.
> We are building an app/library that adds functionality to `django`. So in my opinion, it don't think it would be great to restrict something that `django` allows.

Do you know the reason behind allowing this? Are you sure it's not a bug? In the end Django is not a Holy Book.

Reply:
> Do you know the reason behind allowing this? Are you sure it's not a bug?

Good point, as you can in see the raw SQL queries below, duplication isn't actually passed to the database.

```python
>>> print(Comment.objects.all().order_by('posted', '-posted').query)                        
SELECT "comment_comment"."id", "comment_comment"."user_id", "comment_comment"."email", "comment_comment"."parent_id", "comment_comment"."content_type_id", "comment_comment"."object_id", "comment_comment"."content", "comment_comment"."urlhash", "comment_comment"."posted", "comment_comment"."edited" FROM "comment_comment" ORDER BY "comment_comment"."posted" ASC

>>> print(Comment.objects.all().order_by('-posted', 'posted').query)                        
SELECT "comment_comment"."id", "comment_comment"."user_id", "comment_comment"."email", "comment_comment"."parent_id", "comment_comment"."content_type_id", "comment_comment"."object_id", "comment_comment"."content", "comment_comment"."urlhash", "comment_comment"."posted", "comment_comment"."edited" FROM "comment_comment" ORDER BY "comment_comment"."posted" DESC

# even for more than two values they consider just the first value of the fields.
>>> print(Comment.objects.all().order_by('-posted', 'posted', 'posted').query)              
SELECT "comment_comment"."id", "comment_comment"."user_id", "comment_comment"."email", "comment_comment"."parent_id", "comment_comment"."content_type_id", "comment_comment"."object_id", "comment_comment"."content", "comment_comment"."urlhash", "comment_comment"."posted", "comment_comment"."edited" FROM "comment_comment" ORDER BY "comment_comment"."posted" DESC

>>> print(Comment.objects.all().order_by('-posted', 'posted', '-posted').query)             
SELECT "comment_comment"."id", "comment_comment"."user_id", "comment_comment"."email", "comment_comment"."parent_id", "comment_comment"."content_type_id", "comment_comment"."object_id", "comment_comment"."content", "comment_comment"."urlhash", "comment_comment"."posted", "comment_comment"."edited" FROM "comment_comment" ORDER BY "comment_comment"."posted" DESC
```

I will try to see more in their code and see what more I can gather.